### PR TITLE
[ROCm][Perf] Fuse Kimi-K3 FP8 pre-route projections

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Purpose

Integrate the Kimi-K3 gfx950 pre-route primitives through one AMD-only,
fail-closed ownership boundary:

- prepack routed-down and shared-expert weights after checkpoint loading;
- produce routed inputs, shared-expert output, and FP32 router logits in the
  AITER mixed-precision grid; and
- pass the precomputed shared output through a scoped seam so the generic
  shared-expert lifecycle does not repeat the work.

FP8 is opt-in with `VLLM_ROCM_USE_KIMI_K3_PREROUTE_FP8=1`; the exact BF16 path
is separately opt-in, and FP8 takes precedence when both are enabled.
Unsupported shapes, dtypes, devices, architectures, LoRA configurations, or
AITER installations use the existing implementation. NVIDIA and other models
are unchanged.

Review map:

- `amd/ops/moe_preroute.py`: checked AITER adapter and weight packing;
- `amd/linear.py`: Kimi-K3 lifecycle and dispatch;
- `shared_experts.py`: scoped precomputed-output ownership seam;
- tests: operator contract, fallback, graph replay, and equal-logit behavior.

Depends on ROCm/aiter#4498 and ROCm/aiter#4504.

## Test plan

Tested on 8x MI355X (`gfx950`) with the public Kimi-K3 image and
`moonshotai/Kimi-K3@9f62e4e9`. Kernel heads: AITER #4498 `0d755fe1` and #4504
`3d1208d4`; the command below pins the vLLM base and PR revisions.

Fetch and verify the vLLM source:

```bash
git clone https://github.com/vllm-project/vllm.git vllm
git -C vllm fetch origin refs/pull/50665/head:pr-50665
test "$(git -C vllm rev-parse pr-50665)" = \
  84a9b03c803421804eb3062de43022edce6a8992
git -C vllm checkout --detach 6c91de36897932ba9b5adb11992235a2a789e009
git -C vllm diff 6c91de36897932ba9b5adb11992235a2a789e009..pr-50665 \
  --binary | git -C vllm apply -
git -C vllm diff --check
```

Construct both arms from the public image using the complete verified vLLM and
AITER trees above, including AITER `csrc/` and `hsa/`; do not copy individual
files. Verify `import vllm, vllm._C, aiter` in the GPU container before testing
to prevent a stale or missing overlay.

Focused tests and companion microbenchmarks:

```bash
python -m pytest -q \
  ../aiter/op_tests/flydsl_tests/test_kimi_k3_moe_preroute_bf16.py \
  ../aiter/op_tests/flydsl_tests/test_kimi_k3_moe_preroute_fp8.py \
  tests/model_executor/layers/fused_moe/test_shared_experts.py \
  tests/models/kimi_k3/test_amd_moe_preroute.py
```

Serve each arm with the same exact model and flags; set both pre-route variables
to `0` for control:

```bash
VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_FP4BMM=1 \
AITER_SITUV2_A8W4=1 AITER_BF16_FP8_MOE_BOUND=0 \
VLLM_ROCM_USE_KIMI_K3_PREROUTE_BF16=1 \
VLLM_ROCM_USE_KIMI_K3_PREROUTE_FP8=1 \
vllm serve /model --served-model-name moonshotai/Kimi-K3 \
  --tensor-parallel-size 8 --trust-remote-code --moe-backend auto \
  --gpu-memory-utilization 0.95 --max-num-seqs 128 \
  --max-num-batched-tokens 4096 --max-model-len 1048576 \
  --enable-prefix-caching --kv-cache-dtype fp8 --reasoning-parser kimi_k3
```

After one 8K/128 warmup, run three 8K/1K batch-one trials with seeds 1--3,
temperature zero, and `--ignore-eos`. Run full `lm-eval==0.4.12` GSM8K on both
arms: 1,319 questions, 5-shot, greedy completion, 2,048 generated tokens,
concurrency 128, and seed 42.

## Test results

- Combined focused suite: **21/21 passed**.
- Changed-file pre-commit, `git diff --check`, and DCO passed.

| MI355X microbenchmark | Control | Candidate | Speedup |
| --- | ---: | ---: | ---: |
| BF16 pre-route | 21.9358 us | 17.0914 us | **1.2834x** |
| FP8 routed/shared output | 17.6833 us | 13.8263 us | **1.2790x** |
| FP8 tri-projection | 20.3035 us | 10.6829 us | **1.9006x** |

| TP8 8K/1K median | Control | Candidate | Change |
| --- | ---: | ---: | ---: |
| Decode throughput | 36.0596 tok/s/GPU | 38.2324 tok/s/GPU | **+6.026%** |
| TPOT | 27.7319 ms | 26.1558 ms | **-1.5760 ms** |

Full GSM8K: control **1270/1319**, candidate **1272/1319**, zero invalid or
transport failures; 14 wins/12 losses (`p=0.8450`). This paired run found no
statistically detectable accuracy difference.

The public suite also exposed two exactly tied router logits (experts 111 and
184 at 189.0). The test requires the exact selected top-16 set and a strict
16/17 boundary (174.0 versus 167.0), without imposing undefined tie order.

## Overlap and limits

- Operator tests establish the stated BF16/FP8 numerical contracts; GSM8K is
  the model-quality gate.
- ROCm/aiter#4498 and ROCm/aiter#4504 own the kernels; this PR only owns vLLM lifecycle and
  dispatch. The two paths share one weight lifecycle, so splitting them would
  duplicate the model integration and shared-output ownership seam.

## Tool assistance
OpenAI Codex assisted with implementation, tests, benchmarking, and drafting.

## Long-prefix agentic serving gate (2026-08-02)

The exact user-requested serving sweep was run with official `aiperf==0.11.0`
on 8x MI355X (TP8), FP8 KV cache, prefix caching, and no speculative decoding:

```bash
aiperf profile --model moonshotai/Kimi-K3 --tokenizer moonshotai/Kimi-K3 \
  --tokenizer-trust-remote-code --url http://127.0.0.1:8000 --api-key EMPTY \
  --endpoint-type chat --streaming --use-server-token-count \
  --num-prefix-prompts 8 --prompt-prefix-length 63240 \
  --synthetic-input-tokens-mean 4760 --synthetic-input-tokens-stddev 0 \
  --output-tokens-mean 350 --output-tokens-stddev 0 \
  --extra-inputs ignore_eos:true --extra-inputs min_tokens:350 \
  --extra-inputs max_tokens:350 --warmup-request-count 3 --sweep-type zip \
  --concurrency 1,8,16,24,48 --request-count 5,40,80,120,240 \
  --random-seed 42 --ui simple
```

The parent, BF16 intermediate, and FP8 candidate each completed **485/485**
requests (**1,455/1,455 total**) with byte-identical generated inputs, exactly
350 output tokens per request, and zero errors, cancellations, output-size
mismatches, or endpoint fatal signatures.

| Concurrency | Requests/arm | Output tok/s, parent -> BF16 -> FP8 | FP8 vs parent | Mean ITL, ms, parent -> FP8 | Mean TTFT, ms, parent -> FP8 | Mean E2E, ms, parent -> FP8 |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 5 | 14.12 -> 14.27 -> 14.41 | **+2.07%** | 51.70 -> 50.13 (**-3.02%**) | 6,743 -> 6,785 (+0.62%) | 24,785 -> 24,281 (**-2.03%**) |
| 8 | 40 | 84.02 -> 83.23 -> 84.38 | +0.43% | 78.43 -> 78.02 (-0.53%) | 5,701 -> 5,714 (+0.23%) | 33,074 -> 32,941 (-0.40%) |
| 16 | 80 | 145.22 -> 145.03 -> 145.36 | +0.09% | 85.87 -> 87.01 (+1.33%) | 7,776 -> 7,632 (-1.85%) | 37,744 -> 37,999 (+0.68%) |
| 24 | 120 | 191.53 -> 191.62 -> 191.44 | -0.05% | 97.16 -> 94.91 (-2.32%) | 9,078 -> 10,069 (+10.92%) | 42,988 -> 43,191 (+0.47%) |
| 48 | 240 | 296.30 -> 296.46 -> 296.51 | +0.07% | 124.75 -> 125.15 (+0.32%) | 12,306 -> 12,154 (-1.24%) | 55,845 -> 55,830 (-0.03%) |

The FP8 path provides a clear concurrency-one benefit; c8-c48 is neutral.
Both optimized support predicates require activation shape `(1, 7168)`, so
larger scheduler batches are expected to use the existing fallback. This
dispatch explanation is inferred from the source contract; no dispatch counter
was collected. TTFT is queueing-sensitive, so the mixed TTFT deltas—especially
c24—are reported without an improvement claim. Each cell is one profile trial
and therefore a point estimate.

Provenance: public image
`vllm/vllm-openai-rocm:kimi-k3@sha256:5aa7e626ff73672f5ca7aae46754570488c23d33ca1ac90756a1d2d1a3fe099b`,
model `moonshotai/Kimi-K3@9f62e4e9`, and random seed 42. The four runtime files
at current PR head `84a9b03c803421804eb3062de43022edce6a8992` are SHA-256-identical to the
measured candidate. The seven runtime files at AITER #4498 head `0d755fe1` and
#4504 head `3d1208d4` are likewise byte-identical to the measured AITER stack.

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
Nobody is going to easily review a FULLY vibe coded PR.
